Mlem/App/Utility/Extensions/Views/View Modifiers/MarkReadOnScroll.swift

```diff
@@ -0,0 +1,46 @@
+//
+//  MarkReadOnScroll.swift
+//  Mlem
+//
+//  Created by Eric Andrews on 2024-08-28.
+//
+
+import Foundation
+import MlemMiddleware
+import SwiftUI
+
+private struct MarkReadOnScroll: ViewModifier {
+    @Setting(\.markReadOnScroll) var markReadOnScroll
+    @Setting(\.postSize) var postSize
+    
+    var index: Int
+    var post: any Post2Providing
+    var postFeedLoader: CorePostFeedLoader
+    
+    func body(content: Content) -> some View {
+        content
+            .task {
+                do {
+                    if markReadOnScroll, try await post.api.batchMarkReadEnabled {
+                        postFeedLoader.stageForMarkRead(before: index, offset: postSize.markReadOffset)
+                    }
+                } catch {
+                    handleError(error)
+                }
+            }
+            .onDisappear {
+                if markReadOnScroll {
+                    post.markReadIfStaged()
+                }
+            }
```

Comment:
I wonder if there's a way to avoid storing `readStaged` on every post - we should try to keep UI logic out of MlemMiddleware if possible. Perhaps `PostGridView` could store a `bottomAppearedPostIndex: Int` value, which is passed as a binding to this `MarkReadOnScroll` modifier? Then the above logic could look something like:

```swift
.task {
    if markReadOnScroll, try await post.api.batchMarkReadEnabled {
        bottomAppearedPostIndex = max(index, bottomAppearedPostIndex)
    }
}
.onDisappear {
    if markReadOnScroll, index <= (bottomAppearedPostIndex - postSize.markReadOffset) {
        post.markRead()
    }
}
```

It should be possible to avoid touching MlemMiddleware at all by using something like that.

I may have overlooked a reason why that might not work. I'm happy to leave this as OOS for now if we want to prioritise shipping a working version ASAP

Reply:
> a bottomAppearedPostIndex: Int value, which is passed as a binding to this MarkReadOnScroll modifier?

Or instead of a binding, a wrapper class to avoid causing a view update

Reply:
Yeah, that's much cleaner--the logic is a bit ugly but it's better than putting it in middleware.